Fabric.Identity.API/Services/ClaimsService.cs

```diff
@@ -48,6 +54,62 @@ public ClaimsResult GenerateClaimsForIdentity(AuthenticateInfo info, Authorizati
             return result;
         }
 
+        /// <summary>
+        /// Gets the subject id based on the identity provider.
+        /// If provider is Azure AD, then look it up from the claims response from Azure
+        /// Otherwise, the user will have the subject ID to use for everything else
+        /// </summary>
+        /// <param name="claimInformation">Claims information from GenerateClaimsForIdentity</param>
+        /// <param name="user">User generated from logging into Identity</param>
+        /// <returns>The subject id</returns>
+        public string GetEffectiveSubjectId(ClaimsResult claimInformation, User user)
+        {
+            CheckWhetherArgumentIsNull(user, nameof(user));
+            CheckWhetherArgumentIsNull(claimInformation, nameof(claimInformation));
+
+            string subjectId = null;
+            if (this.IsExternalTokenAzureAD(claimInformation.SchemeItem))

```

Comment:
Not necessary now, but we may want to factor the `GetEffectiveSubjectId` and `GetEffectiveUserId` methods out into interfaces, and then encapsulate the AD and Azure AD specific logic in their own classes. Then as we add new IdPs, this class doesn't have to change. It's just a matter of adding a new IdP-specific class for extracting this type of information.

Reply:
I was thinking about that as I put these new logic in, these if blocks are going to be a pain if we dont encapsulate them.  Perhaps a plugin model.  But totally agree.